### ISSUE TITLE
Turn this into a Monorepo

### ISSUE THREAD

**Issue by probonopd** [AppImage/AppImageKit#931]
Did not know that it is actually a thing, but according to
https://en.wikipedia.org/wiki/Monorepo

> Google, Facebook, Microsoft, Uber, and Twitter all employ very large monorepos.

The advantages (and disadvantages) are given in the article too.

We could put most of https://github.com/AppImage into this repository to benefit from the advantages given in the Wikipedia article. It's pretty much how I like things - self-standing, all in one, no entangled dependencies between projects.

**Comment by TheAssassin** [comment]
That wikipedia article is so wrong in so many ways. It reads like an "industry paper", lots of unproved information that sounds like facts but is scientifically worthless. There's just a lot of wild claims in there that I could prove wrong one by one with ease. We already explained a lot of times why a properly split source code infrastructure is beneficial for our project. I don't want to repeat myself once again. This is not a suitable source code architecture, at least for us, but I also consider it wrong for almost any other project I can think of.

**Comment by azubieta** [comment]
@probonopd in the past I used one of those "mono-repos". The one that contains the whole Android project. It's a huge thing, about 300 projects live there and it weighs about 80 Gb at that time. But was nothing else but a collection of small projects. I guess that is served the purpose of having EVERYTHING under control while building the OS.
This having everything under control is relevant to our project but, we need small projects just to ease the development. 
If you provide a few uses cases were having such repo is required we can schedule making it using a similar approach to that Google does on Android. 
I would like to mention that this is not a priority for our project right now.

**Comment by probonopd** [comment]
Reopened because I still like the idea of having a monorepo. It would remove the need for relying on transfer.sh to get build artefacts from e.g., a branch of libappimage and/or appimagetool into linuxdeployqt.

**Comment by TheAssassin** [comment]
> Reopened because I still like the idea of having a monorepo.

You're pretty alone with that position. The facts say "don't do it".

> It would remove the need for relying on transfer.sh (which is often broken) to get build artefacts from e.g., a branch of libappimage and/or appimagetool into linuxdeployqt.

Wrong. Monorepos are not about build artifacts, they're about code. What you suggest is force-updating all the files just for random experiments. You should probably read about what monorepos are in more detail. Git submodules or other version-based solutions are much more suitable to pin specific versions of code used in projects.

**Comment by TheAssassin** [comment]
Android is by the way a bad example (at least since 2016, when I built my first LineageOS image). Virtually any real Android distribution uses this `repo` wrapper tool for version control systems which fetches things from hundreds of different single repositories... So, ***no monorepo there***.

**Comment by probonopd** [comment]
> Monorepos are not about build artifacts, they're about code. 

Let me explain:

By scattering code around in different repositories, when something breaks and we find out only a couple of days/weeks later, it is difficult to restore the exact combination of everything when it was still working. When having a monorepo, then all you do is a bisect of one repo, roll back to the last known good state, and have _everything_ back in working order.

I am working from the assumption that all I can care about is master, and when something breaks master it's best to get notified immediately so that it can be fixed. What is not ideal is that today you can e.g., change something in libappimage and/or AppImageKit, and linuxdeployqt is affected only when it is built next time. By having a monorepo, everything would be built on each git push, and errors would be seen instantly. At least that's the theory.

**Comment by probonopd** [comment]
https://github.com/AppImage/appimaged/issues/80 is a point in case.

**Comment by shoogle** [comment]
I wouldn't recommend this. A better alternative would be to include all the other repos as submodules inside one big repo, like Qt does with their [qt5](https://github.com/qt/qt5) "super module" repo. That makes it clear that development is expected to happen in the sub repos rather than the super repo, thus keeping the history nice and clean and separate.

You can run a travis build in the super repo every few days in the super repo to check that everything plays nicely together, or even have it triggered automatically by pushes to the sub repos.

**Comment by probonopd** [comment]
That might be one option indeed, yes.

Advantages over a monorepo
* ?

Disadvantages compared to a monorepo
* Issues are still scattered across many subprojects
* The git submodules need to be manually(?) updated every day or so
* We don't get immediate continuous builds for each git commit
* Pull requests don't get continuous builds of everything

**Comment by shoogle** [comment]
Travis CI can push commits to GitHub so submodules could be updated automatically. The other "disadvantages" are actually *features* of the system, which could be regarded as pros or cons depending on your point of view.

- Most issues only impact one of the subprojects, so there is no value in shoving them all in one tracker, and in fact it makes things much neater if you keep them all separate. When issues affect multiple projects you can link to them from issues in each tracker.
- PR builds are faster because only the current subproject is built
- Contributors only have to deal with one subproject at a time
    - They are not overwhelmed by a huge volume of code
    - Their local builds cannot be broken by changes to other subprojects

The price of this, as you say, is that you don't immediately spot changes that break other projects. Of course, those kinds of changes should really be caught by tests in the libappimage repo, but some problems are bound to slip by undetected. It's a matter of weighing up the pros and cons and deciding whether it makes more sense for the projects to be together or separate.

You can even have a two-way system where commits to a sub repo trigger the super repo to update its submodules and run a build, and if the super build is successful then it triggers the sub repo to update its internal copy of libappimage automatically. You'd need to be careful about submodule recursion and infinite build loops, but it should be possible.

**Comment by shoogle** [comment]
However, having read AppImage/appimaged#80 I think that any kind of monorepo or superepo-with-submodules is overkill for what you are trying to do here. Everything depends on `libappimage`, so you just need to have builds in `libappimage` trigger builds in the other projects too. If the triggered builds are successful then they can create a new commit to update that repo's internal copy of `libappimage` automatically, or post a message on IRC if the build fails so that you can sort it out manually.

**Comment by TheAssassin** [comment]
> Everything depends on libappimage, so you just need to have builds in libappimage trigger builds in the other projects too. If the triggered builds are successful then they can create a new commit to update that repo's internal copy of libappimage automatically, or post a message on IRC if the build fails so that you can sort it out manually.

That's what I suggested long time ago: do proper continuous integration on a system dedicated for the purpose that allows for that kind of flexibility.

With the current system, the only real solution to do half automated CI (that is, some test-update-build-and-merge workflow) would be to auto-send PRs to update libappimage or any other internal dependency to all repositories, have things build there, then update.

There's a lot of problems with that with "just" GitHub/Travis. First, the amount of PRs would be ridiculously high, keeping an overview over that would be pretty pointless. Creating ten PRs for every push to libappimage is stupid IMO, and way overkill.

Next, we don't have any real testing set up. The tests in libappimage don't even cover all lines yet, and only test some functionality. There's no complete unit test infrastructure that you'd need for real CI.

Now, that wouldn't be so much of a big deal if there were good unit tests in all the other projects _that depend on libappimage_ which you can rely on with an acceptable probability (>90% maybe). But there's not a single test in any other project. You wouldn't necessarily need entire unit test infra, but at least some functionality testing that simulate the "normal workflows" with the tools and validate the results.
The only "test" (which *is not a test*, seriously...) would be @probonopd downloading stuff, trying it for a while and then giving approval. That's not reliable, as none of these "tests" covers all the functionality, only a small percentage.

This workflow only makes things worse, not better. 

A monorepo doesn't solve either of these issues, it is just annoying. Right now, we have a update-when-necessary policy. You need a reason like "new feature that we want/need" or "bug in code, fix available" to ever update a submodule. That worked _very well_ over time, using known good versions of dependencies. @probonopd himself also often tells he doesn't see a reason to update things all the time. Why do it here? I don't see the benefits of continuous updates very much. Update when necessary, that keeps the workload low and when there's an actual update, it's 1 minute of work to bring a submodule up to date. Not really a big deal, not even when you have to actually clone first and then update...

As @shoogle wrote, it might be an option to just do this "build them all at once" in a separate repo, but still, I don't see why there's a need to constantly keep everything updated. I don't want to lose our fast builds (I've been working very hard on speeding up things). I actually want to make them faster, not slower. 5-10 minutes *plus the starting time of Travis or whatever* is waaaay too long, CI must be responsive, not slow and bulky.

In fact, at some point, there shall be some prebuilt versions of all our libraries and external dependencies we must build ourselves that can be used to speed up things. A monorepo _the way @probonopd suggests_ also prohibits that workflow completely.

Using a monorepo is not an option. It's just a crazy idea to lose the independence between the code again. Keeping up voting for it IMO is like a kid crying "but I want it so much!", there's no convincing argument towards it, but so many valid arguments against. All the pro arguments quickly got a con reply.

**Comment by shoogle** [comment]
I agree with pretty much everything @TheAssassin said. It's not actually necessary to create PRs for each submodule update as you can just do it locally (or on Travis) and push straight into master. This avoids creating tons of PRs but it still pollutes the history with lots of "update submodules" commits. That problem can be reduced by:

- Don't update submodules after every build
    - just do it every X builds, or Y days after previous update
- Update the submodules in all automated builds, *but don't push the update to the repo*
    - This way you always build against the latest version, but you must update the checked-in submodule manually to avoid unnecessary commits

It's not necessary to have `libappimage` builds wait for the other builds to complete. I'd just carry on pushing commits to `libappimage` and then revert/fix them later if it ends up breaking other projects, as you do currently. The other builds would just be a way to catch errors sooner rather than avoid them entirely.

**Comment by TheAssassin** [comment]
@shoogle the core issue here was the lack of good automated testing, which would allow for such a "update regularly" workflow.
